include/votca/xtp/matrixfreeoperator.h

```diff
@@ -166,27 +166,25 @@ struct generic_product_impl<votca::xtp::MatrixFreeOperator, Mtype, DenseShape,
       Index blocks = op.size() / blocksize;
       // this uses the fact that all our matrices are symmetric, i.e we can
       // reuse half the blocks
-      std::vector<Eigen::MatrixXd> thread_wiseresult(
-          votca::xtp::OPENMP::getMaxThreads(),
-          Eigen::MatrixXd::Zero(dst.rows(), dst.cols()));
-#pragma omp parallel for schedule(guided)
+      Eigen::MatrixXd result = Eigen::MatrixXd::Zero(dst.rows(), dst.cols());
+
+#pragma omp declare reduction (+: Eigen::MatrixXd: omp_out=omp_out+omp_in)\
```

Comment:
I don't fully understand this snippet, could you please tell me how it works?

Reply:
`omp_out=omp_out+omp_in` is the communicative operation to perform and `omp_priv` is the "zero" element, right?

Reply:
if the `reduction` for `MatrixXd` has been declared in `eigen.h` why don't you use it here?

Reply:
this is in the eigen::internal namespace, so I had to do it again.